### PR TITLE
Add instance types to logging, use them in the battery logging messages

### PR DESCRIPTION
**Pull request by WickedShell**
This follows onto #9414 by moving the logging of battery messages to use an instance field. If you only use one battery monitor (far and away the most common setup) this won't have any impact. If you use multiple monitors you will need GCS tooling to catch up to this to graph fields usefully.

For GCS Authors: There is now a unit type of `I` which means that it is an instance message. The field should be used to split what instance on the vehicle it corresponds to. Thus all BAT messages with instance 0 is the first battery monitor, instance 1 is the second monitor. This allows the GCS to then limit graphing to a per battery instance, while not having to update predefined graphs everytime we add more batteries to the logs. (More importantly we were consuming the 8 bit log ID space much to quickly which meant we were going to have problems adding more messages, this should let us claw a lot back).

The UI will need some thoughts, in MAVExplorer this could be done `graph BAT.Volt@0 BAT.Volt@1` to graph the first and second batteries against each other. (GCS's should also feel free to extend this as they desire, for example including the concept of if the index is not specified then graph the first instance only, or potentially for tools which make lists the way APMPlanner2/Mission Planner do they could just make a heading in the UI for BAT0 BAT1 BAT2 automatically as reasonable).

Final other notes: This is a lot more logging then just current, so I took the time to rename those parts, and I also directly inlined the instance logging function. On a function like this which is only used from one caller I find it much more frustrating to have it elsewhere in the code, particularly when all it actually was being done was just called in a `for` loop. This inlining saves 28 bytes, and should run faster since it allows us to skip fetching the singleton for every battery instance.

EDIT: It should also be note a number of MAVLink messages need instance related graphing such as `BATTERY_STATUS` already so this should be considered as well.

### CODE REVIEW

**Comment by peterbarker** [comment]
Discussion results in `BAT[1].Volt` syntax instead of the @ syntax

**Comment by peterbarker** [comment]
For the GCS authors - please note that integers are only one example of what an index might be.

For things like named-value-floats the logical index field is a string.

File: libraries/DataFlash/LogFile.cpp [hunk 1]
```diff
@@ -1394,66 +1394,47 @@ void DataFlash_Class::Log_Write_AttitudeView(AP_AHRS_View &ahrs, const Vector3f
...
+            LOG_PACKET_HEADER_INIT(LOG_BATTERY_MSG),
             time_us             : time_us,
-            voltage             : battery.voltage(battery_instance)
+            instance            : i,
+            voltage             : battery.voltage(i),
+            voltage_resting     : battery.voltage_resting_estimate(i),
+            current_amps        : battery.current_amps(i),
+            current_total       : battery.consumed_mah(i),
+            consumed_wh         : battery.consumed_wh(i),
+            temperature         : (int16_t)(has_temp ? (temp * 100) : 0),
```
Comment (peterbarker) [inline libraries/DataFlash/LogFile.cpp]: Maybe int16_t-min or int16_t-max?

File: libraries/DataFlash/LogStructure.h [hunk 2]
```diff
@@ -91,6 +91,7 @@ const struct UnitStructure log_Units[] = {
     { 'G', "Gauss" },         // Gauss is not an SI unit, but 1 tesla = 10000 gauss so a simple replacement is not possible here
     { 'h', "degheading" },    // 0.? to 359.?
     { 'i', "A.s" },           // Ampere second
+    { 'I', "Instance" },      // Message contains instance fields which should be graphed seperately
```
Comment (peterbarker) [inline libraries/DataFlash/LogStructure.h]: Comment needs help.  I don't know what, though.

Comment (WickedShell) [reply libraries/DataFlash/LogStructure.h]: Agreed, I was unsure where to go with it and it shows :)

File: libraries/DataFlash/LogFile.cpp (hunk 1 above)
Comment (WickedShell) [reply libraries/DataFlash/LogFile.cpp]: This was prompted by users not liking their logs showing a large value before the battery was detected. (Same way they wanted the per cell voltage to not be a large value before connected). It's a cave in to the users over sanity for log tools.

**Comment by WickedShell** [comment]
Different version of this was done (and merged) in a PR by Peter.

**Pull request closed without merging**